### PR TITLE
[connector/spanmetrics]Extend connector to count events for all signals

### PR DESCRIPTION
**Pull request by lahsivjar**
The PR extends the current `spanmetricsv2` connector to count all signal types. The motivation behind adding this feature is to have a single component that can produce metrics from all signal types for common counting needs. The counting can be either simple counts for all event types or histogram-type distribution for span metrics using span duration. Normal counting will count the number of spans, datapoints or logrecords as per the provided config.

Below are the changes performed in this PR. Note that the PR includes some breaking changes. Since the component is not in any release cycle, the changes are listed for ease of review only and are not a concern.

* Extends the component to consume metric and logs in addition to spans.
* [**Breaking**] Renames the previous `counter` type to `sum_and_count` type. This aggregation type is required for span destination metric which is indexed as 2 counters rather than `aggregate_metric_double`.
* [**Breaking**] Introduces a new flavored `counter` type which can count the number of events if enabled. As of now, counting can be performed for spans, datapoints (metrics), and log records.
* [**Breaking**] Drops the default configuration option when the component is enabled without any config. Now, this mode will be an error prompting user to provide at least one aggregation.

Apologies for the big PR, I could have split this into a couple of small PRs but it would have still required a big refactoring PR so I chose to do it in 3 PRs, following work will be follow-ups:

* Rename the component from `spanmetricsv2` to `signaltometrics`.
* Introduce OTTL based filtering conditions to aid in counting.
* `ephemeral_resource_id` config will be removed as per https://github.com/elastic/opentelemetry-collector-components/issues/159

Related to: https://github.com/elastic/opentelemetry-dev/issues/372

### CODE REVIEW

**Review by axw — COMMENTED** [review]
@lahsivjar I'm not sure about this configuration, and I'm a bit concerned about starting this way.

Right now spans are our only _required_ use case for metrics with both a count and a value, but I'd like to see this generalised to application-specific attributes (see other comment).

What do you think about this:
 - keep `spanmetricsv2` as it is for now, i.e. summarising span duration
 - introduce `metricsconnector` which for now will only count log records, metric data points, trace spans (profile samples?)
 - later on we would extend `metricsconnector` with the ability to calculate metrics from count+value. Not sure what the config for that looks like yet, which is part of my hesitation to build a unified connector right now

(Then once we have done all of those things we should open a discussion about donation.)

File: connector/spanmetricsconnectorv2/config/config.go [hunk 1]
```diff
@@ -55,35 +52,37 @@ const (
...
+//
+// For spans, the connector can count the number of span events or aggregate them
+// based on their durations in histograms, summaries, or as 2 sum metrics.
+//
+// For all other event types, the connector can count the number of events.
+//
+// All metrics produced by the connector are in delta temporality.
 type Config struct {
-	Spans []MetricInfo `mapstructure:"spans"`
+	Spans      []SpanMetricInfo `mapstructure:"spans"`
```
[the remark this answers is not in the document]
Comment (lahsivjar) [reply connector/spanmetricsconnectorv2/config/config.go:64]: > keep spanmetricsv2 as it is for now, i.e. summarising span duration
introduce metricsconnector which for now will only count log records, metric data points, trace spans (profile samples?)

Won't metricsconnector will be same as upstream countconnector then? We will have to build the same resource attribute handling into count connector basically - I am not so sure about this. Copying another connector just to add resource attribute handling feels a bit off. Maybe we can do something with https://github.com/elastic/opentelemetry-collector-components/issues/159 by using transform processor but I am not sure if ottl can access telemetry settings.

> deriving request-level metrics

This and the following examples gave me some good points to think about. I didn't think we would need a way to aggregate other metric types into histograms for example. Thinking out loud, histograms & summaries will basically aggregate something like `count` and `value`. After deliberating on this, I came up with the below set of configurations:

```
histogram|summary|sum_and_count:
  explicit: {}
  # count will default to counting the number of events if the statement is empty, default will not use OTTL
  countStatement: <ottl_statement_to_extract_count_as_int>
  # no defaults for value, even span duration MUST define OTTL
  valueStatement: <ottl_statement_to_extract_count_as_double>
count:
  # count will default to counting number of events if statement is empty, default will not use OTTL
  countStatement: <ottl_statement_to_extract_count_as_int>
```
Basically, the above config would be available for all `spans`, `datapoints`, and `logs`. WDYT? Does it make sense?

[the remark this answers is not in the document]
Comment (felixbarny) [reply connector/spanmetricsconnectorv2/config/config.go:64]: Good discussion! I like the idea of having a singe connector that can turn any numeric value of any signal into a histogram or counter. Ideally, it would even be able to turn a gauge metric into a histogram. It seems like valueStatement would be able to do that. I hope that OTTL is efficient enough for that use case.

I'm less sure about `sum_and_count`. It seems to be very specific to how we modeled the metric `span.destination.service.response_time.sum.us` and `span.destination.service.response_time.count`. If we were do do this from scratch again, I think we would have used a histogram/aggregate_metric_double instead.

I'm wondering if that could be expressed by two different `count` configurations. Or maybe we need something like a `sum` that produces a gauge (or counter?) metric but  doesn't have a `countStatement` but a `valueStatement`.

We should make sure that we can evolve the connector to pre-aggregate the metrics for a whole batch. For example, if there are 10 spans with the same set of labels in a single batch, we should be able to emit a single histogram (instead of 10) that aggregate the spans. I think that will be essential for performance. Even if we choose to not implement this behavior right away, we should be able to evolve the connector in this direction, without having to introduce breaking changes to the way it is configured.

[the remark this answers is not in the document]
Comment (lahsivjar) [reply connector/spanmetricsconnectorv2/config/config.go:64]: > Ideally, it would even be able to turn a gauge metric into a histogram. It seems like valueStatement would be able to do that. I hope that OTTL is efficient enough for that use case.

Interesting idea, theoretically this should be possible, and glossing over the [datapoint paths](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/contexts/ottldatapoint/README.md), it should work with the above config.

> I'm less sure about sum_and_count. It seems to be very specific to how we modeled the metric span.destination.service.response_time.sum.us and span.destination.service.response_time.count. If we were do do this from scratch again, I think we would have used a histogram/aggregate_metric_double instead.

Agreed! `sum_and_count` is definitely not going to fly and it is a temporary Elastic-only thing. I intend to remove it but I haven't been able to come up with any ideas yet.

> I'm wondering if that could be expressed by two different count configurations. Or maybe we need something like a sum that produces a gauge (or counter?) metric but doesn't have a countStatement but a valueStatement.

This sounds like a good idea, since both values are represented by [sum metric type](https://opentelemetry.io/docs/specs/otel/metrics/data-model/#sums), we could do a `sum` config with a `valueStatement`. I am working on a quick PoC of how the code will look like to see how the config comes out to be, will check this too.

> We should make sure that we can evolve the connector to pre-aggregate the metrics for a whole batch. For example, if there are 10 spans with the same set of labels in a single batch, we should be able to emit a single histogram (instead of 10) that aggregate the spans.

We already do this, check out the test cases like [this one for histograms](https://github.com/elastic/opentelemetry-collector-components/blob/main/connector/spanmetricsconnectorv2/testdata/with_custom_histogram_configs/output.yaml).

[the remark this answers is not in the document]
Comment (lahsivjar) [reply connector/spanmetricsconnectorv2/config/config.go:64]: I am having a difficult time wrapping my head around the [OTTL grammar](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/LANGUAGE.md). As per the grammar,

```
There must be a single Editor Invocation in each OTTL statement
```
These editors modify telemetry, so, there seems to be no way to get data from telemetry using OTTL - at least not using the currently [available editors](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg/ottl/ottlfuncs#editors). One option would be to create a `get` editor to just return the value and editors do support returning value but while this would be inline with the OTTL grammar in terms of specs, I wonder if it will be inline with the intention behind the grammar.

[the remark this answers is not in the document]
Comment (lahsivjar) [reply connector/spanmetricsconnectorv2/config/config.go:64]: I have a working PoC for this: https://github.com/elastic/opentelemetry-collector-components/pull/163 The code only works for spans as of now and requires a lot of refactoring but the config gives an idea of how it might look like:

```
spanmetricsv2:
  spans:
    - name: http.trace.span.duration
      description: Span duration for HTTP spans
      attributes:
        - key: http.response.status_code
      statements:
        value: get(Milliseconds(end_time - start_time))
      histogram:
        explicit: {}
        exponential: {}
    - name: db.trace.span.duration
      description: Span duration for DB spans
      attributes:
        - key: db.system
      statements:
        value: get(Milliseconds(end_time - start_time))
      histogram:
        explicit: {}
        exponential: {}
```

[the remark this answers is not in the document]
Comment (felixbarny) [reply connector/spanmetricsconnectorv2/config/config.go:64]: This looks great!

I wonder if we even need something like a get editor. Can't we just do something like
```yml
statements:
  value: Milliseconds(end_time - start_time)
```

[the remark this answers is not in the document]
Comment (lahsivjar) [reply connector/spanmetricsconnectorv2/config/config.go:64]: This was what I first thought but we can't. As per the OTTL grammar, we must have a single editor in the OTTL statement and `Milliseconds` (plus other such funcs) are converters. The parsing of OTTL statement fails when you don't have a editor in the statement 😞 (it will work for conditions, but we need statement). I have also created [this PR](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35618) upstream to get some ideas/feedbacks.

[the remark this answers is not in the document]
Comment (felixbarny) [reply connector/spanmetricsconnectorv2/config/config.go:64]: For the `.count`/`.sum.us`, maybe we don't need to have different type for it. Could we just have two counter metrics where the `.count` metric has a statement like `value: 1`, and the `.sum.us` metric has something like `Milliseconds(end_time - start_time)`?

Could we also enable the use of conditions to do stuff like count the number of logs where the level is error? Maybe `value: 1 WHERE severity_text: 'ERROR'` or `condition: WHERE severity_text: 'ERROR', value: 1`.

[the remark this answers is not in the document]
Comment (felixbarny) [reply connector/spanmetricsconnectorv2/config/config.go:64]: But maybe we can also simplify for now and push out to full OTTL support for later. We could have helpers for the common case where we want to get the span duration or the event count.

Something like

```yml
value:
  source: duration_ns|duration_us|count|ottl
  statement: <ottl statement in case type is ottl>
```

[the remark this answers is not in the document]
Comment (lahsivjar) [reply connector/spanmetricsconnectorv2/config/config.go:64]: I plan to add conditionals too. Currently, I am thinking of utilizing the `WHERE` in the same statement but I haven't tested how that works yet.

> Could we just have two counter metrics where the .count metric has a statement like value: 1, and the .sum.us metric has something like Milliseconds(end_time - start_time)?

+1, this looks like the cleanest solution to me. I plan to keep the count always optional i.e. if no count statement is provided it will take the count of that specific event type.

[the remark this answers is not in the document]
Comment (lahsivjar) [reply connector/spanmetricsconnectorv2/config/config.go:64]: This is my proposal for the configuration:

```yaml
signaltometrics:
  spans|datapoints|logs: # can be extended to metric and spanevent if required in future.
    intStatement: string # ottl statement that must produce a int64 or int64 like value.
    doubleStatement: string # ottl statement that must produce a float64 or float64 like value.
    histogram:
      explicit:
        buckets: []float64 | has_default | optional
      exponential:
        max_size: int64 | has_default | optional
      # For histograms the intStatement will refer to the count and the doubleStatement to the value.
      # doubleStatement MUST be present while intStatement is optional. Config will be validated.
      # If intStatement is not present count will be calculated as the absolute count of events under
      # which it is defined.
    summary:
      # For summary the intStatement will refer to the count and the doubleStatement to the value.
      # doubleStatement MUST be present while intStatement is optional. Config will be validated.
      # If intStatement is not present count will be calculated as the absolute count of events under
      # which it is defined.
    sum:
      # For sum the intStatement will create a sum metric with integer values and doubleStatement a
      # sum metric with double value.
      # For sum exactly one of intStatement OR doubleStatement MUST be present, and will be validated.
    count:
      # count is a special shorthand config for sum with intStatement recording the absolute count of
      # events under which it is defined.
      # This provides an optimum way to calculate count of signals without relying on OTTL statements
      # returning literals and wraps calculation for absolute count.
      # For count NONE of the intStatement OR doubleStatement are required, and will be validated.
```

[the remark this answers is not in the document]
Comment (axw) [reply connector/spanmetricsconnectorv2/config/config.go:64]: >> I wonder if we even need something like a get editor. Can't we just do something like
>
> This was what I first thought but we can't. As per the OTTL grammar, we must have a single editor in the OTTL statement and Milliseconds (plus other such funcs) are converters. The parsing of OTTL statement fails when you don't have a editor in the statement 😞 (it will work for conditions, but we need statement). I have also created this PR upstream to get some ideas/feedbacks.

Should the configuration specify an [OTTL value](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/LANGUAGE.md#values), rather than a statement?

```yaml
signaltometrics:
  spans|datapoints|logs: # can be extended to metric and spanevent if required in future.
    statements:
      int: string # OTTL statement that must produce a int64 or int64 like value.
      double: string # OTTL statement that must produce a float64 or float64 like value.
```

Is it necessary to explicitly specify the type in the field, or could that be implied by the evaluated expression?

```yaml
    histogram:
      explicit:
        buckets: []float64 | has_default | optional
      exponential:
        max_size: int64 | has_default | optional
```

Maybe worth defining these as two separate top-level metric types? e.g. `explicit_histogram` and `exponential_histogram`. (Since they have no common properties.)

```yaml
    count:
      # count is a special shorthand config for sum with intStatement recording the absolute count of
      # events under which it is defined.
      # This provides an optimum way to calculate count of signals without relying on OTTL statements
      # returning literals and wraps calculation for absolute count.
      # For count NONE of the intStatement OR doubleStatement are required, and will be validated.
```

This feels like it's leaking what could be an internal optimisation to the user. IIANM think this would be equivalent to a `sum` with a statement/value with a literal value of 1 (or, I suppose, the representative count?). Could we identify such a value, and optimise it internally if needed?

[the remark this answers is not in the document]
Comment (lahsivjar) [reply connector/spanmetricsconnectorv2/config/config.go:64]: > Should the configuration specify an [OTTL value](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/LANGUAGE.md#values), rather than a statement?

Discussed this IRL, since the current OTTL package doesn't have an option to parse ottl Value (with converters), we decided to do this internally(converter ottl.Value+converters -> editor using the temporary `get` editor) while keeping the outside configuration specify ottl.Value.

> Maybe worth defining these as two separate top-level metric types? e.g. explicit_histogram and exponential_histogram. (Since they have no common properties.)

Sounds good, I have been dragging this along since I first thought it would be a good idea. Happy to discard this now.

> This feels like it's leaking what could be an internal optimisation to the user. IIANM think this would be equivalent to a sum with a statement/value with a literal value of 1 (or, I suppose, the representative count?). Could we identify such a value, and optimise it internally if needed?

Discussed this IRL too, it would be a bit complicated to handle this internally. I will create a PR with the current decision and discuss this point in the new PR.

**Comment by lahsivjar** [comment]
Closing this as decision as been reached to do this differently (see the comments above).
